sklearn/model_selection/_search.py

```diff
@@ -552,59 +593,140 @@ def _fit(self, X, y, labels, parameter_iterable):
+        self.n_fits_ = out.shape[0]
+
+        self.n_candidates_ = self.n_fits_ // n_splits
+        self.n_parameters_ = len(self._parameter_names)
+
+        res_shape = (self.n_candidates_,)
+
+        # Lets not initite this everytime and reuse the same array.
+        test_scores = np.empty((n_splits,), dtype=np.float64)
+        test_sample_counts = np.empty((n_splits,), dtype=np.float64)
+
+        search_results = dict()
+
+        for param in self._parameter_names:
+            # One column to record the values of each parameter
+            search_results["param_%s" % param] = (
+                np.ma.masked_all(res_shape, dtype=object))
+
+        # Make a column for each split of each metric
+        for split_i in range(n_splits):
+            search_results["test_split_%s" % split_i] = (
+                np.empty(res_shape, dtype=np.float64))
+        search_results["test_mean"] = np.empty(res_shape, dtype=np.float64)
+        search_results["test_std"] = np.empty(res_shape, dtype=np.float64)
+        search_results["test_rank"] = np.empty(res_shape, dtype=int)
+
+        for candidate_i in range(self.n_candidates_):
+            grid_start = candidate_i * n_splits
+            grid_end = grid_start + n_splits
+
+            test_scores = out[grid_start:grid_end, 0]
+            test_sample_counts = out[grid_start:grid_end, 1]
 
-        scores = list()
-        grid_scores = list()
-        for grid_start in range(0, n_fits, n_splits):
-            n_test_samples = 0
-            score = 0
-            all_scores = []
-            for this_score, this_n_test_samples, _, parameters in \
-                    out[grid_start:grid_start + n_splits]:
-                all_scores.append(this_score)
-                if self.iid:
-                    this_score *= this_n_test_samples
-                    n_test_samples += this_n_test_samples
-                score += this_score
             if self.iid:
-                score /= float(n_test_samples)
+                mean_score = (test_scores * test_sample_counts).sum()
+                mean_score /= test_sample_counts.sum()
             else:
-                score /= float(n_splits)
-            scores.append((score, parameters))
-            # TODO: shall we also store the test_fold_sizes?
-            grid_scores.append(_CVScoreTuple(
-                parameters,
-                score,
-                np.array(all_scores)))
-        # Store the computed scores
-        self.grid_scores_ = grid_scores
+                mean_score = test_scores.mean()
+
+            for split_i in range(n_splits):
+                search_results["test_split_%s" % split_i][candidate_i] = (
+                    test_scores[split_i])
+
+            # Store the mean score and the parameters for this fit
+            search_results["test_mean"][candidate_i] = mean_score
+            search_results["test_std"][candidate_i] = test_scores.std()
+
+        candidate_dicts = list(parameter_iterable)
```

Comment:
If this is truly a more-or-less arbitrary iterable, it's possible the iterable has been exhausted at this point, and the list will come out empty. If you need to make a list of it, you need to do it before fitting. Or else, use `return_parameters`=True